### PR TITLE
torch.flip and torch.flip{lr, ud}: Half support for CPU and BFloat16 support for CPU & CUDA

### LINKED ISSUE #49889: float16 tensors in CPU should be able to be flipped
## To Reproduce

        >>> a = torch.tensor([1,2])
        >>> a.flip(-1)
        tensor([2, 1])
        >>> a = torch.tensor([1,2]).cuda()
        >>> a.flip(-1)
        tensor([2, 1], device='cuda:0')
        >>> a = torch.tensor([1,2]).half().cuda()
        >>> a.flip(-1)
        tensor([2., 1.], device='cuda:0', dtype=torch.float16)
        >>> a = torch.tensor([1,2]).half()
        >>> a.flip(-1)
        Traceback (most recent call last):
            File "<input>", line 1, in <module>
        RuntimeError: "flip_cpu" not implemented for 'Half'

## Expected behavior

 torch.tensor([1,2]).half() should be flipped

## Environment
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.5 LTS (x86_64)
GCC version: (Ubuntu 6.4.0-17ubuntu1~16.04) 6.4.0 20180424
Clang version: 3.8.0-2ubuntu4 (tags/RELEASE_380/final)
CMake version: version 3.5.1

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 440.64.00
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.1
[pip3] torchvision==0.8.2
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               10.2.89              hfd86e86_1    defaults
[conda] mkl                       2020.2                      256    defaults
[conda] mkl-service               2.3.0            py37he8ac12f_0    defaults
[conda] mkl_fft                   1.2.0            py37h23d657b_0    defaults
[conda] mkl_random                1.1.1            py37h0573a6f_0    defaults
[conda] numpy                     1.19.2           py37h54aff64_0    defaults
[conda] numpy-base                1.19.2           py37hfa32c7d_0    defaults
[conda] pytorch                   1.7.1           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchvision               0.8.2                py37_cu102    pytorch

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #49889

Also adds BFloat16 support for CPU and CUDA

### CODE REVIEW

**Review by kshitij12345 — COMMENTED** [review]
Haven't updated the testing logic in test_shape_ops.py as they mainly check the behavior of `flip` but not the supported types.

https://github.com/pytorch/pytorch/blob/bc4ff7ba0506190524d598be1bbe8112f3e09935/test/test_shape_ops.py#L339-L445

**Comment by ngimel** [comment]
Can you please add bfloat16 support too?

**Comment by kshitij12345** [comment]
Sure.

File: test/test_shape_ops.py [hunk 1]
```diff
@@ -386,11 +386,11 @@ def test_flip(self, device):
         self.assertEqual(flip1_result, data.flip(1))
 
         # test empty tensor, should just return an empty tensor of the same shape
-        data = torch.tensor([])
+        data = torch.tensor([]).to(device)
```
Comment (mruberry) [inline test/test_shape_ops.py:389]: Thank you for fixing this. A better pattern would be:

```
data = torch.tensor((), device=device)
```

Then the tensor wouldn't have to be created and copied to the device.

File: test/test_shape_ops.py [hunk 2]
```diff
@@ -386,11 +386,11 @@ def test_flip(self, device):
...
         self.assertEqual(data, data.flip(0))
 
         # test bool tensor
-        a = torch.tensor([False, True])
+        a = torch.tensor([False, True]).to(device)
```
Comment (mruberry) [inline test/test_shape_ops.py:393]: Also a good fix for the existing test, but like the above the device can be passed to `torch.tensor()`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -546,6 +547,28 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
...
+    test_cases = (
+        make_tensor((S, M, S), device, dtype, low=None, high=None, requires_grad=requires_grad),
+        make_tensor((S, 0, M), device, dtype, low=None, high=None, requires_grad=requires_grad)
+    )
+
+    dims = [[0, 1, 2], [0], [0, 2], [-1]]
+    out = []
+    for dim in dims:
+        for case in test_cases:
+            random.shuffle(dim)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:560]: Let's make a few tweaks here:

- rename `test_cases` to `tensors`
- add back the case where dim is `()`
- make `dims` a tuple of tuples
- rename `out` to `samples`
- iterate over a product of `tensors` and `dims`
- remove the randomization by killing the `random.shuffle`
- I think with the above changes you won't have to `tuple(dim)`

**Review by mruberry — APPROVED** [review]
Nice work, @kshitij12345. I made a few inline recommendations that should be easy to adopt. Just ping me when they're done.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:560]: Interestingly,

```python
>>> a = torch.randn(3,2,0,2)
>>> torch.flip(a, dims=())
tensor([], size=(3, 2, 0, 2))
>>> b = a.cuda()
>>> torch.flip(a, dims=())
tensor([], size=(3, 2, 0, 2))
>>> torch.flip(b, dims=())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: flip dims size out of range, got flip dims size=0
```

Should I just add a test in `test_shape_ops` which verifies this behavior i.e. works on CPU but assertRaises on CUDA?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:560]: Why doesn't it work on CUDA?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:560]: That is, there's probably a bug here:

https://github.com/pytorch/pytorch/blob/b824fc4de22c0e278ec57669b1b82c2d8e1f7394/aten/src/ATen/native/cuda/TensorTransformations.cu#L71

Filing an issue, testing for the current behavior + citing the issue, and fixing this in a future PR would be cool.

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:560]: Sure. Thanks for the reference!

**Comment by kshitij12345** [comment]
@mruberry Have addressed the review. PTAL :)

File: test/test_shape_ops.py [hunk 4]
```diff
@@ -378,21 +378,31 @@ def test_flip(self, device):
             self.assertEqual(size, list(data.flip(ds).size()))
 
         # test rectangular case
-        data = torch.tensor([1, 2, 3, 4, 5, 6]).view(2, 3).to(device)
-        flip0_result = torch.tensor([[4, 5, 6], [1, 2, 3]]).to(device)
-        flip1_result = torch.tensor([[3, 2, 1], [6, 5, 4]]).to(device)
+        data = torch.tensor([1, 2, 3, 4, 5, 6], device=device).view(2, 3)
```
Comment (ngimel) [inline test/test_shape_ops.py:381]: why do we need separate `test_flip`? Are there any tests here that are not covered by OpInfo tests?

Comment (mruberry) [reply test/test_shape_ops.py:381]: There are, unfortunately. Maybe in the future OpInfos will cover everything here.

Comment (kshitij12345) [reply test/test_shape_ops.py:381]: The Generic OpInfo tests do not verify the expected output of the operator against some reference.
That happens for UnaryUfuncInfo and others but not the generic one.

Thus the above tests are not covered.

Generic OpInfo tests cover basic behaviour of the operator like
* Raising an error for unsupported dtypes.
* Not raising an error for supported dtypes.
* Tests operators for consistency between JIT and eager
* gradcheck and gradgradcheck
* Verifying that out= output is same compared to when out is not passed (for applicable operators)

Reference: https://github.com/pytorch/pytorch/blob/master/test/test_ops.py

**Pull request closed without merging**